src/logging.h

```diff
@@ -77,6 +77,8 @@ namespace BCLog {
         Info,      // Default
         Warning,
         Error,
+        Critical,
+        Fatal,
```

Comment:
The medium post describes five levels with the same names as ours currently: error, warn, info, debug, trace; the other six links all describe syslog's behaviour, which was designed for unix systems in the 1980s. Logging systems we actually use include python's (which has 5 default levels), leveldb (which does not have levels, so we treat it all as debug), and libevent (which has 4 levels). So, no, 6-8 levels is not standard, and for a custom subsystem we should be doing what's best for us, not necessarily what's standard in any case.

The flowchart from the stackoverflow link might be helpful, though: https://stackoverflow.com/a/64806781/84932 For us I think it would look more like:

```mermaid
flowchart TD
    start-->|No|dbg
    start{{Is this critical information?}}-->|Yes|crit

    dbg-->|No|logtrace(Trace + Category)
    dbg{{Is this moderate volume?}}-->|Yes|logdebug(Debug + Category)

    crit{{Is there a problem?}}-->|No|loginfo(Info)
    crit-->|Yes|warnerr

    warnerr{{Do we need to abort?}}-->|No|logwarn(Warning/Critical)
    warnerr-->|Yes|logfatal(Error/Fatal)
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30347#discussion_r1659004164

> Meta:
> 
> It's not clear why people are 👍 on that comment.

I kind of like the mystery of the reactions. But I 👍 AJ's chart because I think it is a good recommendation for how to choose log levels in new code, even if it doesn't really tell you what log levels mean in existing code. I also broadly agree with having few log levels and using few log levels. I think by a large margin most log statements should be using LogDebug.

Reply:
> Your [flowchart](https://github.com/bitcoin/bitcoin/pull/30347#discussion_r1657886163) is a good idealization, but it does not describe the realities of current code, which is using logging for ordinary error reporting, not just the most critical errors.

The current code mostly hasn't migrated to the multiple severity levels from the original LogPrintf/LogPrint split. It would be nice to spend time working on that, but we're instead constantly bikeshedding the API instead and treating that as a blocker for any other logging changes.

Reply:
It's not clear to me that there's any reason to separate the current `LogWarning` / `LogError` levels; those existed prior to #28318 and I documented them in that PR as best I understood them. However, there's not much benefit in a log that says "hey this error is about to cause the node to stop" -- you already get that information by seeing "Shutdown: in progress..." immediately following. So it could make sense to merge Warning/Error into:

 * - `LogAlert(fmt, params...)` should be used in place of `LogInfo` for (potentially) severe problems that require attention from the node admin (e.g., system time appears to be wrong, unknown soft fork appears to have activated, insufficient storage space).

which would reduce the flowchart to:

```mermaid
flowchart TD
    start-->|No|dbg
    start{{Is this critical information?}}-->|Yes|crit

    dbg-->|No|logtrace(Trace + Category)
    dbg{{Is this moderate volume?}}-->|Yes|logdebug(Debug + Category)

    crit{{Is there a problem?}}-->|No|loginfo(Info)
    crit-->|Yes|logfatal(Alert)
```

Reply:
By overwrought, jonatack means `LogError("CHECKSUM ERROR ...")`, while his preferred non-overwrought api from #25203 looks like `LogPrintLevel(BCLog::NET, BCLog::Level::Error, "CHECKSUM ERROR ...")`.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30347#discussion_r1659096893

I like your idea to collapse the log levels further and I am curious about ways you think we can make progress towards that goal.

> However, there's not much benefit in a log that says "hey this error is about to cause the node to stop" -- you already get that information by seeing "Shutdown: in progress..." immediately following.

If we don't need to have a dedicated level for fatal errors, it suggests a possible alternative to this PR, implemented in #30361. It would solve the issue that was reported and might get closer to the long term outcome you outlined.





 So it could make sense to merge Warning/Error into:
> 
> * `LogAlert(fmt, params...)` should be used in place of `LogInfo` for (potentially) severe problems that require attention from the node admin (e.g., system time appears to be wrong, unknown soft fork appears to have activated, insufficient storage space).
> 
> which would reduce the flowchart to:
> 
> ```mermaid
> flowchart TD
>     start-->|No|dbg
>     start{{Is this critical information?}}-->|Yes|crit
> 
>     dbg-->|No|logtrace(Trace + Category)
>     dbg{{Is this moderate volume?}}-->|Yes|logdebug(Debug + Category)
> 
>     crit{{Is there a problem?}}-->|No|loginfo(Info)
>     crit-->|Yes|logfatal(Alert)
> ```